### PR TITLE
check-for-changes: performance improvements + wait for settle

### PR DESCRIPTION
**Pull request by NorseGaud**
# Description

See: https://github.com/docker-mailserver/docker-mailserver/issues/2098

This PR will:

- Have `check-for-changes.sh` wait for changes to "settle" (checksum comparison has to not have changed over two different checks with 5 second gaps between) before modifying and restarting anything
- Added backgrounding of processes and using wait throughout instead of sleeps for more efficiency
- Removed the while loop from addmailuser due to check-for-changes.sh logic changes

This is, in my opinion, critically needed for an API to be performing changes -- and lots of them all at once! See https://github.com/docker-mailserver/docker-mailserver-admin/issues/3 for more discussion about this.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change that does improve existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by NorseGaud** [comment]
Forgot to draft this... again... Sorry... It's still a WIP

**Comment by NorseGaud** [comment]
It looks like both the CI and my local machine are getting the following (running with new verbose flag):

```
✗ begin 186 checking tls: cipher list - rsa intermediate
   (from function `assert_success' in file test/test_helper/bats-assert/src/assert.bash, line 114,
    from function `collect_cipherlist_data' in file test/security_tls_cipherlists.bats, line 154,
    from function `check_ports' in file test/security_tls_cipherlists.bats, line 86,
    in test file test/security_tls_cipherlists.bats, line 47)
     `check_ports 'rsa' 'intermediate'' failed
   $ docker run -d --name tls_test_cipherlists --volume /Users/norsegaud/docker-mailserver/test/duplicate_configs/security_tls_cipherlists.bats/:/tmp/docker-mailserver/ --volume /Users/norsegaud/docker-mailserver/test/test-files/ssl/example.test/:/config/ssl/:ro --env DMS_DEBUG=0 --env ENABLE_POP3=1 --env SSL_TYPE=manual --env SSL_CERT_PATH=/config/ssl/cert.rsa.pem --env SSL_KEY_PATH=/config/ssl/key.rsa.pem --env TLS_LEVEL=intermediate --network test-network --network-alias example.test --hostname mail.example.test --tty mailserver-testing:ci
     f7d78f74fc3eedaa1277f63d9b7a2d06f22fbae7705f75b842b56bed71fce40b
   [ TASKLOG ]  mail.example.test is up and running
   $ docker run --rm --user 501:20 --network test-network --volume /Users/norsegaud/docker-mailserver/test/test-files/ssl/example.test/:/config/ssl/:ro --volume /var/folders/vb/pjyhgj4s491_k0t025_j56y00000gn/T//results/rsa/intermediate/:/output --workdir /output drwetter/testssl.sh:3.1dev --quiet --file /config/ssl/testssl.txt --mode parallel --overwrite --preference
     standard_init_linux.go:228: exec user process caused: permission denied
   
   -- command failed --
   status : 1
   output : standard_init_linux.go:228: exec user process caused: permission denied
   --
```

It seems like `--user 501:20` is the issue as it runs without it. The comment says 

>  # `--user "<uid>:<gid>"` is a workaround: Avoids `permission denied` write errors for json output, uses `id` to match user uid & gid.

@polarathene, I'm going to see if anything else fails after removing it but I wanted to see if it's ok if it's removed (as long as everything else passes).

**Comment by polarathene** [comment]
> --user 501:20 --network test-network --volume /Users/norsegaud/

That is odd looking uid and gid values, and with that volume is this Windows/WSL? That context seems to be missing in the CI tests (due to lack of verbose flag I guess?)

But this other volume looks more of an issue:

> --volume /var/folders/vb/pjyhgj4s491_k0t025_j56y00000gn/T//results/rsa/intermediate/:/output

I'm not familiar with this mount, but that `T//results` looks a bit broken path wise? I don't think it should have two `/` in a row like that?

**Comment by polarathene** [comment]
I believe whatever changes have been introduced have broken the functionality of `${BATS_TMPDIR}` providing a valid path prefix here:

https://github.com/docker-mailserver/docker-mailserver/blob/f33374023f94be9a6daf2093f3cf0bdbf8d1d7aa/test/security_tls_cipherlists.bats#L22-L23

**Comment by NorseGaud** [comment]
Thanks @polarathene , appreciate your insite. So, I rolled bats back to 1.3.0 and it works. It looks like 1.4.0 introduced a change to BATS_TMPDIR: https://github.com/bats-core/bats-core/pull/410#discussion_r575864351

How the hell this didn't fail in https://github.com/docker-mailserver/docker-mailserver/pull/2095 is beyond me... Might be good to have @casperklein and @georglauterbach take a look if they get a few minutes. I'm unsure about how the CI is all set up TBH.

**Comment by polarathene** [comment]
> How the hell this didn't fail in #2095 is beyond me...

I'd have to refresh my memory but I recall PRs can behave differently depending on what you're doing. Secrets aren't usually available for example, and in one case I think you end up with the default branch sha/ref depending on what you're doing. 

Perhaps the tests didn't get run on that PR with the updated submodule? And only after merging to the default branch does that apply? I know when setting up CI for the docs there were a few gotchas I ran into like this :\

**Comment by NorseGaud** [comment]
Please ignore the bats submodule for now. I'm using that to gain visibility into the tests and will revert it soon.

**Comment by georglauterbach** [comment]
Is this WIP?

**Comment by NorseGaud** [comment]
@georglauterbach , no it's ready for review

**Review by georglauterbach — APPROVED** [review]
LGTM 👍🏼

**Comment by georglauterbach** [comment]
@NorseGaud when you push yourself, my review will be dismissed I think. But @docker-mailserver/maintainers can update the branch via the webUI which _should not_ dismiss it (if that is somehow important) :D

File: test/open_dkim.bats
Comment (williamdes) [inline test/open_dkim.bats:24]:
```suggestion
		-v "$PWD/test/test-files":/tmp/docker-mailserver-test:ro \
```

Comment (NorseGaud) [reply test/open_dkim.bats:24]: We use $(pwd) in a ton of other tests. What's the reason to use $PWD here instead?

Comment (williamdes) [reply test/open_dkim.bats:24]: Oh no, it was for me more practical then calling a binary
But if tests already use it then its perfect that way ;)

Comment (NorseGaud) [reply test/open_dkim.bats:24]: Roger that -- Within the tests/* directory I see 60 occurrences of `$(pwd)` :)

**Comment by georglauterbach** [comment]
LGTM 👍 

@casperklein or @wernerfred feel free to merge this PR if you approve it :)

**Review by georglauterbach — APPROVED** [review]

**Comment by georglauterbach** [comment]
I will go ahead and merge this :)

**Comment by casperklein** [comment]
I hadn't time to review this yet. Otherwise I had approved it..

**Comment by wernerfred** [comment]
Dito ⌛

File: target/supervisor/conf.d/supervisor-app.conf
Comment (NorseGaud) [reply target/supervisor/conf.d/supervisor-app.conf:111]: https://github.com/docker-mailserver/docker-mailserver/pull/2137

Comment (NorseGaud) [reply target/supervisor/conf.d/supervisor-app.conf:111]: If you find any others, let me know and I'll PR them. Maybe we could add a little comment in the supervisor file about the dependency? Or maybe add a test?

Comment (wernerfred) [reply target/supervisor/conf.d/supervisor-app.conf:111]: Add a test seems appropriate

Comment (casperklein) [reply target/supervisor/conf.d/supervisor-app.conf:111]: > If you find any others, let me know and I'll PR them

Yes, I think, I will fully review this PR until next week end. That was just something, that immediately kept my attention.

Comment (casperklein) [reply target/supervisor/conf.d/supervisor-app.conf:111]: A comment and/or test is good. Regarding a test, I don't know how to test this behaviour properly. Quick test would just be, to grep and see if "stopwaitsecs=55" is present or not.

File: target/scripts/postfix-wrapper.sh
Comment (casperklein) [inline target/scripts/postfix-wrapper.sh:24]: I think the `sleep` here was, because `service postfix start` exit (0) successfully pretty fast, however postfix needs some time to come up completely.

```
service postfix start &
wait
```

is equal to just `service postfix start` ?

I am not sure if it's related, but I read on a issue/PR this week, about our CI behaving "flacky" again. Maybe this is the reason?

File: target/scripts/check-for-changes.sh
Comment (casperklein) [inline target/scripts/check-for-changes.sh:57]: I know it was there before, but do some know what the `--` are for? I couldn't find something in the man page.

Comment (NorseGaud) [reply target/scripts/check-for-changes.sh:57]: > A double-dash in a shell command signals the end of options and disables further option processing.

Comment (georglauterbach) [reply target/scripts/check-for-changes.sh:57]: Commonly used for example when you're trying to `grep` something that starts with a double dash, you'd write

``` BASH
grep -E -- "--to-grep" $FILE
```

Comment (casperklein) [reply target/scripts/check-for-changes.sh:57]: I knew that. I was wondering, if in that case, it has another purpose. Because it seems to me unnecessary here.

File: target/scripts/postfix-wrapper.sh
Comment (NorseGaud) [reply target/scripts/postfix-wrapper.sh:24]: Good points -- I'll play around with it and see what I find.

File: target/scripts/check-for-changes.sh
Comment (NorseGaud) [reply target/scripts/check-for-changes.sh:57]: I never use it personally :)

File: target/scripts/postfix-wrapper.sh
Comment (casperklein) [inline target/scripts/postfix-wrapper.sh:30]: Inspired by your other changes, I wonder if this one liner would work too:

`wait $(</var/spool/postfix/pid/master.pid)`

Comment (casperklein) [reply target/scripts/postfix-wrapper.sh:30]: Unfortunately, `wait` only works on child processes of the current shell:

```
# wait $(</var/spool/postfix/pid/master.pid)
bash: wait: pid 3108 is not a child of this shell 
```

https://stackoverflow.com/a/41613532/568737 suggests `tail` instead (which works):

`tail -f --pid="$(</var/spool/postfix/pid/master.pid)" /dev/null`

We could replace the while loop with that.

Comment (NorseGaud) [reply target/scripts/postfix-wrapper.sh:30]: Sounds good -- including in https://github.com/docker-mailserver/docker-mailserver/pull/2134

Comment (casperklein) [reply target/scripts/postfix-wrapper.sh:30]: I am unsure. This could lead to problems on MacOS? I don't have a mac to test, but the stackoverflow posts suggest this only for linux.
On the other hand, we don't support MacOS officially. What do you think @docker-mailserver/maintainers?

Comment (NorseGaud) [reply target/scripts/postfix-wrapper.sh:30]: The code doesn't run on mac, it'll run inside of the linux contianer. Shouldn't be an issue.

Comment (casperklein) [reply target/scripts/postfix-wrapper.sh:30]: You are absolutely right.

**Pull request merged**
